### PR TITLE
timer form

### PR DESCRIPTION
**Pull request by rStrzelczyk98**

### CODE REVIEW

File: src/app/service/timer-service.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  it('should call addTimer()', () => {
+    spyOn(service, 'addTimer');
+    service.addTimer({
+      label: null,
+      hours: null,
+      minutes: null,
+      seconds: null,
+    });
+    expect(service.addTimer).toHaveBeenCalled();
```
Comment (RdN) [inline src/app/service/timer-service.service.spec.ts:31]: Doesn't make sense - you called it on line 25 :). I rather would get timers and check the length of the array - to check if it was added.

File: src/app/service/timer-service.service.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  private timers: Timer[] = [];
+
+  constructor() {}
+
+  addTimer(timer: Timer) {
+    this.timers.push(timer);
+  }
+
+  getTimers(): Observable<Timer[]> {
+    return of(this.timers);
```
Comment (RdN) [inline src/app/service/timer-service.service.ts:24]: I would keep it in a stream - not just converting [] into the stream. So I would keep it in Behaviour Subject, then push (next(timer)) and use the scan to do the reducer.

File: src/app/timer-form/timer-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  selector: 'app-timer-form',
+  templateUrl: './timer-form.component.html',
+  styleUrls: ['./timer-form.component.scss'],
+})
+export class TimerFormComponent {
+  timerForm!: FormGroup;
+  timers$!: Observable<Timer[]>;
+
+  constructor(private timerService: TimerService) {
+    this.timerForm = new FormGroup(
```
Comment (RdN) [inline src/app/timer-form/timer-form.component.ts:22]: I would use form builder.

File: src/app/timer-form/timer-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      {
+        label: new FormControl(null),
+        hours: new FormControl(null, [
+          this.typeValidator(),
+          this.minValue(),
```
Comment (RdN) [inline src/app/timer-form/timer-form.component.ts:27]: Why not build in NG validators?

**Review by RdN — CHANGES_REQUESTED** [review]

File: src/app/service/timer-service.service.spec.ts (hunk 1 above)
Comment (rStrzelczyk98) [reply src/app/service/timer-service.service.spec.ts]: changed

File: src/app/service/timer-service.service.ts (hunk 2 above)
Comment (rStrzelczyk98) [reply src/app/service/timer-service.service.ts]: Changed to behaviorSubject

File: src/app/timer-form/timer-form.component.ts (hunk 3 above)
Comment (rStrzelczyk98) [reply src/app/timer-form/timer-form.component.ts]: done

File: src/app/timer-form/timer-form.component.ts (hunk 4 above)
Comment (rStrzelczyk98) [reply src/app/timer-form/timer-form.component.ts]: min, max replaced with built-in validators

File: src/app/service/timer-service.service.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  private timers$ = new BehaviorSubject<Timer[]>([]);
+
+  constructor() {}
+
+  getTimers(): Observable<Timer[]> {
+    return this.timers$.asObservable();
+  }
+
+  addTimer(timer: Timer) {
+    this.timers$.next([...this.timers$.value, timer]);
```
Comment (RdN) [inline src/app/service/timer-service.service.ts:24]: Can we do this.timers$.next(timer); and use scan operator to reduce it?

**Review by RdN — CHANGES_REQUESTED** [review]

File: src/app/service/timer-service.service.ts (hunk 5 above)
Comment (rStrzelczyk98) [reply src/app/service/timer-service.service.ts:24]: Throws error: Argument of type 'Timer' is not assignable to parameter of type 'Timer[]'

Comment (RdN) [reply src/app/service/timer-service.service.ts:24]: BEcause it should be type of Timer and reducer should return Timer[]

**Review by RdN — CHANGES_REQUESTED** [review]

**Pull request merged**